inexorgame/inexor-core
# Refactor server/client + shrink monster headers

### what we do with this PR:
* refactor the whole sourcebase to be more modular. (a lot of headers become directly includeable)
* make all of our gigantic common headers a bit smaller
   * tools.hpp even gets completely shrinked
* remove the listenserver
   * this was a functionality for offline modes to run without a server exposed to others.
   * including a lot of extra rules and complexity for them
   * this step was necessary to clean up the design of the client/server model
* remove remaining functionality regarding singleplayer campaigns
   * see point above: were pure multiplayer.
   * if campaigns come back, they will come back as multiplayer campaigns
* refactor the weird gamemode code
   * its still not perfectly in shape, but #339 should be fullfilled
   * (and any system beyond that should probably be more generically designed)

#320 (refactor the gameserver code) is in reach (could be done in this PR, but no need imo)
and even #170 is pretty near.

### Is it working?

Its working for e.g. ffa or teamplay, but some other modes are not fully working yet.  
The following bugs are to solve:

- [x] fix the capture/ctf/bomberman network issues
- [x] fix the mastermode issue
    * if you disconnect with mastermode locked and reconnect you can not claim master
- [x] add PRIV_LOCAL for connections via localhost
- [x] reintroduce localconnect
- [x] Flex: always start a server with inexor
- [x] fix the logger (for multiple instances)
- [x] fix gamespeed issues
- [x] fix the malformatted timestamps
- [x] test the screenshot functionality
- ~~test the demo mode~~ (on the hackathon)
- ~~fix the userinfo about the players being wrongly placed~~ (deferred, probably caused somewhen else)
- [x] fix missing logging of the server

afterwards we could
- [x] enable the server RPC #431 
- [x] rename the inexor binaries #430 

### severity?
This does touch a lot of code. probably best to merge it as soon as the issues are resolved.

### How to test?
build and install as usual (see wiki).
afterwards `npm start` in the flex directory (to start the client)
afterwards/or before that click the server.bat (or the server script for unix)
ingame write `/connect` into the console.
thats it.
